lib/Agrammon/DB/Dataset.pm6

```diff
@@ -542,4 +515,74 @@ class Agrammon::DB::Dataset does Agrammon::DB {
         }
     }
 
+    method store-branch-data(@vars, Str $instance, %options, @fractions) {
+        my $dataset-name = $!name;
+
+        my @branch-variables;
+        # Get variable ids and names
+        self.with-db: -> $db {
+            my $username = $!user.username;
+            @branch-variables = $db.query(q:to/SQL/, $username, $dataset-name, |@vars, $instance).hashes;
+            SELECT data_id, data_var
+                  FROM data_new
+                 WHERE data_dataset=dataset_name2id($1,$2)
+                   AND data_var IN ($3,$4)
+                   AND data_instance = $5
+                   ORDER BY data_id
+            SQL
+        }
+
+        for @branch-variables -> %var {
+            my $var-id   = %var<data_id>;
+            my $var-name = %var<data_var>;
+
+            my @options = %options{$var-name}.map(*.subst(' ', '_', :g) );
+            self.with-db: -> $db {
+                my $ret = $db.query(q:to/SQL/, $var-id, @fractions[*;*], @options);
+                    INSERT INTO branches (branches_var, branches_data, branches_options)
+                                  VALUES ($1, $2, $3)
+                    ON CONFLICT (branches_var)
+                    DO
+                        UPDATE SET branches_data    = EXCLUDED.branches_data,
+                                   branches_options = EXCLUDED.branches_options
```

Comment:
Ah, so nice this logic can be done in the database. :-)

Reply:
Yes, I like it. Not quite standardised in all common DBs (PostgreSQL and SQLite both have it, so that's good enough for what we usually use). Found a few articles about potential problems, but at least for Agrammon I don't think they are of concern.